### PR TITLE
[wip] Refactor carmen-cache

### PR DESCRIPTION
**Pull request by apendleton**
This PR reflects the work-in-progress refactor effort described in #115 .

### CODE REVIEW

File: src/rocksdbcache.cpp [hunk 1]
```diff
@@ -0,0 +1,650 @@
...
+    return _genericget<RocksDBCache>(info);
+}
+
+/**
+ * lists the keys in the RocksDBCache object
+ *
+ * @name list
+ * @memberof RocksDBCache
+ * @param {String} id
+ * @returns {Set} Set of keys/ids
```
Comment (apendleton) [inline src/rocksdbcache.cpp]: Actually an array, not a set

File: src/rocksdbcache.cpp [hunk 2]
```diff
@@ -0,0 +1,650 @@
...
+    baton->method = method;
+    baton->callback.Reset(callback.As<Function>());
+    baton->request.data = baton;
+    uv_queue_work(uv_default_loop(), &baton->request, mergeQueue, static_cast<uv_after_work_cb>(mergeAfter));
+    info.GetReturnValue().Set(Nan::Undefined());
+    return;
+}
+
+/**
+ * Constructor - Create RocksDBCache object which keeps phrases in memory for indexing reference
```
[the remark this answers is not in the document]
Comment (apendleton) [reply src/rocksdbcache.cpp]: Again though, I think the important thing if this is where we're putting what the class is for, is, similar to the memory one: key/value store from key/languages -> grids

File: src/cpp_util.cpp [hunk 3]
```diff
@@ -0,0 +1,141 @@
+
+#include "cpp_util.hpp"
+
+namespace carmen {
+
+// derives grid values
```
Comment (apendleton) [inline src/cpp_util.cpp]: Covers are a representation of `(relev, score, x, y, feature_id)` packed into a single 64-bit integer. This function converts from the packed integer into the full structure where you can access each property.

File: src/cpp_util.cpp [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    // external values after initialization.
+    cover.idx = 0;
+    cover.mask = 0;
+    cover.tmpid = 0;
+    cover.distance = 0;
+
+    return cover;
+}
+
+// Compute distance between coordinates
```
Comment (apendleton) [inline src/cpp_util.cpp]: This isn't a distance calculation. It converts from the ZXY coordinates of the tile that contains a proximity point to a ZXY at the appropriate zoom level necessary for a given coalesce operation (which may examine grids at multiple zoom levels that might not be the same as the zoom level used to specify the proximity location).

File: src/cpp_util.cpp [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    assert(pMid_d <= static_cast<double>(std::numeric_limits<unsigned>::max()));
+    assert(pMid_d >= static_cast<double>(std::numeric_limits<unsigned>::min()));
+    signed pMid = static_cast<signed>(pMid_d);
+    zxy.x = static_cast<unsigned>((static_cast<signed>(x) * zMult) + pMid);
+    zxy.y = static_cast<unsigned>((static_cast<signed>(y) * zMult) + pMid);
+    return zxy;
+}
+
+
+// Compute distance between coords + zoom multiplier
```
Comment (apendleton) [inline src/cpp_util.cpp]: Again, doesn't calculate a distance, but rather, calculates a ZXY for appropriate for a given coalesce operation out of the supplied bbox ZXY coordinates.

File: src/cpp_util.cpp [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    // Since distance is in tiles we calculate scoredist by converting the miles into
+    // a tile unit value at the appropriate zoom first.
+    //
+    // 32 tiles is about 40 miles at z14, use this as our mile <=> tile conversion.
+    scoredist = ((radius * (32.0 / 40.0)) / _pow(1.5, 14 - static_cast<int>(zoom))) / distance;
+
+    return score > scoredist ? score : scoredist;
+}
+
+// database lookup
```
Comment (apendleton) [inline src/cpp_util.cpp]: This doesn't look anything up, it just opens the file so it's ready for lookups later. We do this when a carmen instance starts. This one is read-write, so we mostly use it for indexing.

File: src/cpp_util.cpp [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+rocksdb::Status OpenDB(const rocksdb::Options& options, const std::string& name, std::unique_ptr<rocksdb::DB>& dbptr) {
+    rocksdb::DB* db;
+    rocksdb::Status status = rocksdb::DB::Open(options, name, &db);
+    dbptr = std::unique_ptr<rocksdb::DB>(db);
+    return status;
+}
+
+// read only database lookup
```
Comment (apendleton) [inline src/cpp_util.cpp]: The same, but read-only (used at runtime).

File: src/normalizationcache.cpp [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+ *
+ * @name get
+ * @memberof NormalizationCache
+ * @param {Number} id
+ * @returns {Array}
+ * @example
+ * const cache = require('@mapbox/carmen-cache');
+ * const nc = new cache.NormalizationCache('file.norm.rocksdb', true);
+ *
+ * // for a normalization cache with over the dictionary ['main st', 'main street']
```
Comment (aaaandrea) [inline src/normalizationcache.cpp]: rm `over`

File: src/normalizationcache.cpp [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+    info.GetReturnValue().Set(out);
+    return;
+}
+
+/**
+ * bulk-set the contents of a NormalizationCache to an array of arrays
+ *
+ * @name writeBatch
+ * @memberof NormalizationCache
+ * @returns {Array}
```
Comment (aaaandrea) [inline src/normalizationcache.cpp]: add `@params`

File: src/coalesce.cpp [hunk 10]
```diff
@@ -0,0 +1,719 @@
...
+#include "binding.hpp"
+
+namespace carmen {
+
+using namespace v8;
+
+/**
+ * @typedef PhrasematchSubq
+ * @name PhrasematchSubq
+ * @description The PhrasematchSubq type describes the metadata known about a possible matches to be assessed for stacking by coalesce.
```
Comment (aaaandrea) [inline src/coalesce.cpp]: `about a possible` - rm "a"

Also suggest more description on how this chooses a match candidate.

File: src/coalesce.cpp [hunk 11]
```diff
@@ -0,0 +1,719 @@
...
+/**
+  * @callback coalesceCallback
+  * @param err - error if any, or null if not
+  * @param {CoalesceResult[]} results - the results of the coalesce operation
+  */
+
+/**
+ * @typedef CoalesceResult
+ * @name CoalesceResult
+ * @description The a member of the result set from a coalesce operation.
```
Comment (aaaandrea) [inline src/coalesce.cpp]: something about how it is an object which is that matched a phrase in a carmen-index

File: src/coalesce.cpp (hunk 10 above)
Comment (apendleton) [reply src/coalesce.cpp]: > how this chooses a match candidate

It doesn't. This is just the description of the data format so we can set this as the type of the inputs to coalesce; it's not an actual function. These are supplied by carmen as input to carmen cache, so whatever matching has already occurred on the carmen side (and should presumably be documented there).

File: src/coalesce.cpp (hunk 11 above)
Comment (apendleton) [reply src/coalesce.cpp]: Hm, again, this is just a typedef (to define the format of the output). What the function actually does is described in the docstring for the function. I'm not sure I follow though? A phrase isn't part of the return value, just the properties listed here (x, y, relev, score, id, idx, tmpid, distance, score_dist, and matches_language). The phrase came in as *input*.

File: src/cpp_util.hpp [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+constexpr uint64_t POW2_48 = static_cast<uint64_t>(_pow(2.0, 48));
+constexpr uint64_t POW2_34 = static_cast<uint64_t>(_pow(2.0, 34));
+constexpr uint64_t POW2_28 = static_cast<uint64_t>(_pow(2.0, 28));
+constexpr uint64_t POW2_25 = static_cast<uint64_t>(_pow(2.0, 25));
+constexpr uint64_t POW2_20 = static_cast<uint64_t>(_pow(2.0, 20));
+constexpr uint64_t POW2_14 = static_cast<uint64_t>(_pow(2.0, 14));
+constexpr uint64_t POW2_3 = static_cast<uint64_t>(_pow(2.0, 3));
+constexpr uint64_t POW2_2 = static_cast<uint64_t>(_pow(2.0, 2));
+
+struct PhrasematchSubq {
```
Comment (boblannon) [inline src/cpp_util.hpp]: why not have the typedef doc here?

Comment (apendleton) [reply src/cpp_util.hpp]: We could. I'm inclined not to, though. The typedef describes the way JS objects should be formatted in the in the phrasematches array on the way into coalesce. They come in as JS objects, and then are processed and their contents stored in C++ structs of this type ^. The two types serve the same purpose, but are not the same: aside from being of different types in a technical sense, they also don't have exactly the same fields, either in terms of what they're called or what type they are. Like, `languages` on the JS side is an array of integers, but that gets converted into a bitfield on the C++ side and stored in a property called `langfield`. I sort of think conflating the two is a recipe for confusion. (That said, given that, maybe I should call the JS type something different?)

Also, as a practical matter, at the moment we aren't looking for JSDoc entries in headers, and were also trying to keep the `cpp_util.*` files generally free of JS-isms. Those are both trivially changeable though, so I think they're less compelling arguments than the above.

Comment (boblannon) [reply src/cpp_util.hpp]: > (That said, given that, maybe I should call the JS type something different?)

Would be less confusing if I had `API.md` open and was browsing C++ code...

File: src/cpp_util.hpp [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+
+        s.reserve(sizeof(LANGFIELD_SEPARATOR) + field_length);
+        s.push_back(LANGFIELD_SEPARATOR);
+        s.append(lf_as_char, field_length);
+    } else {
+        s.push_back(LANGFIELD_SEPARATOR);
+    }
+}
+
+// in general, the key format including language field is <key_text><separator character><8-byte langfield>
```
Comment (aarthykc) [inline src/cpp_util.hpp]: 👌

File: src/cpp_util.hpp (hunk 12 above)
Comment (apendleton) [reply src/cpp_util.hpp]: Renamed the JS-accessible version in https://github.com/mapbox/carmen-cache/pull/116/commits/c6d30e51eb2419e41c510de2b2529d93adaf29ab

**Pull request merged**